### PR TITLE
New Ordinals & Testing Framework Refactor

### LINKED ISSUE #74: [TTIOLT] Oltima 2000 - Crashes when navigating menu
**Module Information**
* Identifier: TTIOLT
* Module Name: Oltima 2000
* Version: unknown

**Describe the bug**
Game crashes when navigating initial menu

**To Reproduce**
1. Enter Module
2. Try to move around menu

**Expected behavior**
Ability to "Play" the game and navigate menu

**Screenshots**
Annotation 2020-08-29 104106

```
Unhandled exception. System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'Unknown Exported Function Ordinal in GALGSBL: 22')
   at MBBSEmu.HostProcess.ExportedModules.Galgsbl.Invoke(UInt16 ordinal, Boolean offsetsOnly)
   at MBBSEmu.HostProcess.ExecutionUnits.ExecutionUnit.ExternalFunctionDelegate(UInt16 ordinal, UInt16 functionOrdinal)
   at MBBSEmu.CPU.CpuCore.Tick()
   at MBBSEmu.HostProcess.ExecutionUnits.ExecutionUnit.Execute(IntPtr16 entryPoint, UInt16 channelNumber, Boolean simulateCallFar, Boolean bypassState, Queue`1 initialStackValues, UInt16 initialStackPointer)
   at MBBSEmu.Module.MbbsModule.Execute(IntPtr16 entryPoint, UInt16 channelNumber, Boolean simulateCallFar, Boolean bypassSetState, Queue`1 initialStackValues, UInt16 initialStackPointer)
   at MBBSEmu.HostProcess.MbbsHost.WorkerThread()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ThreadHelper.ThreadStart()
```

**Software Information:**
- OS: Windows Server 2019 Standard
- Telnet Client: NetRunner

**API Report**
{"UniqueIdentifier":"TTIOLT","File":"TTIOLT.DLL","Imports":{"DOSCALLS":[89],"PHAPI":[16],"MAJORBBS":[11,226,43,637,401,77,403,628,520,476,629,377,334,463,543,516,561,599,455,456,492,589,331,113,574,544,624,511,636,560,474,119,485,162,313,53,534,167,565,350,578,571,462,389,566,107,441,650,158,117,522,587,484,625,351,621,459,437,654,659,458,713,559,658,486,317,409,428,601,134,580,475,712,640,639,582,477,505,550,205,225],"GALGSBL":[11,21,30,6,36,19,22,39,60,72,56]}}

### PR DESCRIPTION
**Pull request by enusbaum**
- Added MAJORBBS.155 (daytoday)
- Added MAJORBBS.169 (digalw)
- Added MAJORBBS.127 (cncnum)
- Added GALGSBL.22 (btuinp)
- Refactored Unit Tests for Exported Modules to allow Base Class to test additional Exported Modules
- Small NuGet Package Update

Fixes #74 , #96 , #97 , #72

### CODE REVIEW

File: MBBSEmu.Tests/ExportedModules/Galgsbl/btuimp_Tests.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public class btuimp_Tests : ExportedModuleTestBase
+    {
+        private const int BTUIMP_ORDINAL = 22;
+
+        [Fact]
+        public void btuimp_Test()
+        {
+            //Reset State
+            Reset();
```
Comment (paladine) [inline MBBSEmu.Tests/ExportedModules/Galgsbl/btuimp_Tests.cs:15]: Reset() is only necessary for [InlineData] parameterized tests no?

File: MBBSEmu.Tests/ExportedModules/Galgsbl/btuimp_Tests.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            //Allocate Variables to be Passed In
+            var outputStringPointer = mbbsEmuMemoryCore.AllocateVariable("STRING", 0xFF);
+
+            //Set Session Input Buffer
+            testSessions[0].InputBuffer.Write(Encoding.ASCII.GetBytes("Test Input String"));
+
+            ExecuteApiTest(HostProcess.ExportedModules.Galgsbl.Segment, BTUIMP_ORDINAL, new List<ushort> { 0, outputStringPointer.Offset, outputStringPointer.Segment});
+
+            //Verify Results
+            Assert.Equal("Test Input String\0", Encoding.ASCII.GetString(mbbsEmuMemoryCore.GetString(outputStringPointer)));
```
Comment (paladine) [inline MBBSEmu.Tests/ExportedModules/Galgsbl/btuimp_Tests.cs:26]: let's assert on AX too

**Review by paladine — CHANGES_REQUESTED** [review]

File: MBBSEmu.Tests/ExportedModules/Majorbbs/cncnum_Tests.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+namespace MBBSEmu.Tests.ExportedModules.Majorbbs
+{
+    public class cncnum_Tests : ExportedModuleTestBase
+    {
+        private const int CNCNUM_ORDINAL = 127;
+
+        [Theory]
+        [InlineData("100\0", 0, "100", 4)]
+        [InlineData("-100\0", 0, "-100", 5)]
+        [InlineData("-100abc000\0", 0, "-100", 5)]
```
[the remark this answers is not in the document]
Comment (enusbaum) [reply MBBSEmu.Tests/ExportedModules/Majorbbs/cncnum_Tests.cs:15]: It's the increment on NXTCMD, so it starts at 1. We could probably enhance these tests to also verify the char at NXTCMD to ensure it's at the correct point in the input string.

File: MBBSEmu/HostProcess/ExportedModules/Galgsbl.cs [hunk 4]
```diff
@@ -957,5 +960,36 @@ private void btulok()
...
+            if (!ChannelDictionary.TryGetValue(channelNumber, out var channel))
+            {
+                Registers.AX = ERROR_CHANNEL_NOT_DEFINED;
+                return;
+            }
+
+            //Write to Destination
+            channel.InputBuffer.WriteByte(0); //NULL Terminate
+            Module.Memory.SetArray(destinationPointer, channel.InputBuffer.ToArray());
+            Registers.AX = (ushort)channel.InputBuffer.Length;
```
[the remark this answers is not in the document]
Comment (enusbaum) [reply MBBSEmu/HostProcess/ExportedModules/Galgsbl.cs:989]: You are correct! 

image

File: MBBSEmu.Tests/ExportedModules/Galgsbl/btuimp_Tests.cs (hunk 1 above)
Comment (enusbaum) [reply MBBSEmu.Tests/ExportedModules/Galgsbl/btuimp_Tests.cs:15]: So without the reset the test fails. I think something else is happening in the `Reset()` functions on the CPU, Memory , etc. but I'm not sure what it is exactly. Doesn't hurt having it here atm.

**Pull request merged**